runtime/commonMain/src/kotlinx/serialization/modules/SerialModuleBuilders.kt

```diff
@@ -141,11 +140,48 @@ public class SerializersModuleBuilder internal constructor(@JvmField internal va
         vararg baseClasses: KClass<*>,
         buildAction: PolymorphicModuleBuilder<Any>.() -> Unit = {}
     ) {
-        val builder = PolymorphicModuleBuilder(baseClass as KClass<Any>, null)
+        val builder = PolymorphicModuleBuilder(baseClass as KClass<Any>)
         builder.buildAction()
-        builder.buildTo(impl)
+        builder.buildTo(this)
         for (base in baseClasses) {
-            builder.changeBase(base as KClass<Any>, null).buildTo(impl)
+            builder.changeBase(base as KClass<Any>, null).buildTo(this)
+        }
+    }
+
+    internal fun <T : Any> registerSerializer(
+        forClass: KClass<T>,
+        serializer: KSerializer<T>,
+        allowOverwrite: Boolean = false
+    ) {
+        if (!allowOverwrite && forClass in class2Serializer) throw SerializerAlreadyRegisteredException(forClass)
+        class2Serializer[forClass] = serializer
+    }
+
+    internal fun <Base : Any, Sub : Base> registerPolymorphicSerializer(
+        baseClass: KClass<Base>,
+        concreteClass: KClass<Sub>,
+        concreteSerializer: KSerializer<Sub>,
+        allowOverwrite: Boolean = false
+    ) {
+        // Check for overwrite
+        val name = concreteSerializer.descriptor.serialName
+        val baseClassSerializers = polyBase2Serializers.getOrPut(baseClass, ::hashMapOf)
+        if (!allowOverwrite && concreteClass in baseClassSerializers) throw SerializerAlreadyRegisteredException(
+            baseClass,
+            concreteClass
+        )
+        // Check for serial name conflicts
+        val names = polyBase2NamedSerializers.getOrPut(baseClass, ::hashMapOf)
+        val previous = names[name]
+        if (previous != null) {
```

Comment:
Hmmm, do we have a test that this check works correctly with `allowOverwrite=true` when we try to overwrite the previous serializer by a new one with the same name? Or should we prohibit such operation? 

(btw, aren't we supposed to delete old serial name from the map on overwrite action?)

Reply:
Reworked this, good point, one more old bug caught :)